### DISCUSSION TITLE
Kernel Crashes During YOLOv8 Segmentation Training (Small Dataset, GPU)

### DISCUSSION THREAD

**Discussion by lienpham83** [ultralytics/ultralytics#19236] [Q&A]
I’m training a YOLOv8n segmentation model on a small dataset (28 images, 640x640, 20KB each) using GPU, but my kernel keeps crashing.
I got the segmentation data in the tif format and then I converted it to label text format yolov8 segmentation. When I run trainning, it just generated train and val cache and then kernel dies. I do not know what cause this crahses (my label not correct?) or other issues? How to fix this?

 I attached my zip data below and My system Information:
- OS: Windows 10
- CPU: Intel64 Family 6 Model 143 Stepping 8, GenuineIntel
- RAM: 128 GB
- Python Version: 3.9.18
- PyTorch Version: 2.5.1+cu121
- CUDA Available: True
- CUDA Device Count: 1
- CUDA Version: 12.1
- cuDNN Version: 90100
-GPU 0: NVIDIA RTX A5000
  - Compute Capability: (8, 6)
- Ultralytics Version: 8.3.74 

 My script here:

`>````
def convert_tif_to_yolov8(tif_image_path, output_dir):
    # Load the TIFF annotation image
    image = Image.open(tif_image_path)
    image_array = np.array(image)
    
    # Threshold the image (class 1 is greater than 0, class 0 is 0)
    mask = image_array > 0
    
    # Find contours of the regions of interest (class 1)
    contours, _ = cv2.findContours(mask.astype(np.uint8), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    
    # Get the base name of the input TIFF for naming the output file
    base_name = os.path.basename(tif_image_path).replace('.tif', '')
    annotation_file = os.path.join(output_dir, f"{base_name}.txt")
    
    with open(annotation_file, 'w') as f:
        for contour in contours:
            # Flatten the contour to a list of points
            contour_points = contour.flatten().tolist()
            
            if len(contour_points) > 4:  # Filter out very small regions
                # Convert contour points to YOLO format (normalized x, y coordinates)
                normalized_points = []
                h, w = mask.shape  # height and width of the image
                
                for i in range(0, len(contour_points), 2):
                    # Normalize each point (x, y) by width and height
                    x_norm = contour_points[i] / w
                    y_norm = contour_points[i + 1] / h
                    normalized_points.append(f"{x_norm:.6f} {y_norm:.6f}")
                
                # Write class (1 for the region of interest) and the polygon points
                f.write(f"0 {' '.join(normalized_points)}\n")
    
    print(f"Annotations saved for {tif_image_path} to {annotation_file}")

model = YOLO('yolov8n-seg.pt')  # Load the model
model.to('cuda')
# output_directory = r'C:\delete\Lien\model_train'
# Train the model and save results in the specified directory
model.train(
    data=r'C:\delete\data.yaml',   
    epochs=10,   amp=False,                 
    device='gpu',                    
    batch=1,
    imgsz=640)
```

Below are the zip folder contains orginal image, segementation image, the label format, and the data yml
[data.zip](https://github.com/user-attachments/files/18791100/data.zip)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lienpham83, thank you for sharing your issue and for your interest in Ultralytics 🚀! We understand how frustrating kernel crashes can be. Let's work together to resolve this efficiently. 

Here are some steps and suggestions to get started:

### 🐛 Bug Report
If you suspect this is a bug, could you provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/)? An MRE will help us debug your issue quickly. For example:
- Can you narrow down the issue to a specific step or operation (e.g., loading the dataset, caching, or training)?
- Verify that your dataset annotations are formatted correctly and compatible with YOLOv8. Check the [data format requirements](https://docs.ultralytics.com/datasets/) for your segmentation task.

### Logs and Validation
- Please attach complete training logs if available. They provide vital context for debugging.
- Double-check your `data.yaml` file to confirm the paths and class definitions are correctly specified. For formatting tips, refer to our [Dataset Guide](https://docs.ultralytics.com/datasets/).

### Upgrade and Dependencies
Ensure you're running the latest `ultralytics` version and check that all dependencies are up-to-date:
```bash
pip install -U ultralytics
```

You can also verify your software setup against our verified environments. Make sure you're working in a [Python>=3.8](https://www.python.org/) environment and that your [PyTorch>=1.8](https://pytorch.org/get-started/locally/) installation matches your CUDA version.

### 🚀 Verified Environments
YOLO supports these environments:
- **Notebooks with free GPU**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Desktop Environments**: Ensure your CUDA and cuDNN installations are aligned with your GPU driver version.
You can also test with a [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

### Debugging Dataset
Looking at your code snippet for converting TIFF annotations, ensure all contours are processed correctly in YOLO format. Debugging suggestions:
- Verify the generated `.txt` annotation files to ensure coordinates are normalized (between 0 and 1).
- Test a smaller dataset with the same pipeline to rule out dataset-related formatting issues.

### Upgrade System Libraries
Based on your system specs:
- Your CUDA (`12.1`) and `PyTorch 2.5.1+cu121` versions are compatible, but confirm that all supporting libraries, such as cuDNN (`9.0`), match requirements.
- You can also try running the training on a different verified platform (e.g., Google Colab or Gradient) to isolate whether your setup might be contributing to the issue.

---

### Community Support and Real-Time Help
🧑‍🤝‍🧑 Join our thriving community for more support and solutions:  
- [Discord](https://discord.com/invite/ultralytics) for real-time discussion 🎧.  
- [Discourse](https://community.ultralytics.com/) for in-depth Q&A.  
- [Subreddit](https://reddit.com/r/Ultralytics) for knowledge sharing and ideas.

An Ultralytics engineer will review your issue soon to provide further assistance. Thank you for your patience 😊

**Comment by Y-T-G** [comment]
`device='gpu'` isn't a valid argument. It should be device=0

And don't use `.to('cuda')`

**Reply by lienpham83** [reply]
Thanks Y-T-G. I did your suggestions but it still crashes.

**Reply by glenn-jocher** [reply]
@lienpham83 thanks for the update. Let's verify your dataset format matches YOLO segmentation requirements: each label should be a single line per object with class followed by normalized polygon points. Try training on COCO8-seg first to isolate the issue. If that works, compare your data.yaml structure and label formatting to the COCO8 example. Also check CUDA memory usage during training - with batch=1 and imgsz=640, your A5000 should handle it, but monitor via `nvidia-smi`. For direct debugging, please share a minimal reproducible example.

**Reply by Y-T-G** [reply]
Run it as a script. Not in notebook.

**Reply by Lien-TLL** [reply]
@ glenn-jocher and @Y-T-G: Thank you so much for your suggestions. I did your suggestion, but it still did not work due to the package I installed, and then I reinstalled my package to Ultralytics 8.3.62  Python-3.10.16 torch-2.5.1 CUDA:0 (NVIDIA RTX A5000, 24564MiB), everything works fine.
